dnf/base.py

```diff
@@ -1263,30 +1265,40 @@ def query_for_repo(query):
         return ygh
 
     def _add_comps_trans(self, trans):
-        cnt = 0
+        self._comps_trans += trans
+        return len(trans)
+
+    def _finalize_comps_trans(self):
+        trans = self._comps_trans
         clean_deps = self.conf.clean_requirements_on_remove
-        attr_fn = ((trans.install, self._goal.install),
-                   (trans.install_opt,
+        basearch = self.conf.substitutions['basearch']
+
+        def cond_check(pkg):
+            if not pkg.requires:
+                return True
+            return self.sack.query().installed().filter(name=pkg.requires) or \
+                pkg.requires in (pkg.name for pkg in self._goal._installs) or \
+                pkg.requires in [pkg.name for pkg in trans.install]
+
+        attr_fn = ((trans.install,
                     functools.partial(self._goal.install, optional=True)),
                    (trans.upgrade, self._goal.upgrade),
                    (trans.remove,
                     functools.partial(self._goal.erase, clean_deps=clean_deps)))
 
         for (attr, fn) in attr_fn:
-            for it in attr:
-                if not self.sack.query().filter(name=it):
-                    # a comps item that doesn't refer to anything real
-                    if (attr == trans.install):
-                        self._group_persistor._rollback()
-                        raise dnf.exceptions.MarkingError(it)
+            for pkg in attr:
+                query_args = {'name': pkg.name}
+                if (pkg.basearchonly):
+                    query_args.update({'arch': basearch})
+                if not self.sack.query().filter(**query_args) and \
+                        not cond_check(pkg):
```

Comment:
@mluscon I am not sure but the cond_check is applied for each package. Shall it be applied only to packages of conditional type?